plugins/database/mysql/mysql.go

```diff
@@ -0,0 +1,187 @@
+package mysql
+
+import (
+	"database/sql"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/vault/builtin/logical/database/dbplugin"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/plugins/helper/database/connutil"
+	"github.com/hashicorp/vault/plugins/helper/database/credsutil"
+	"github.com/hashicorp/vault/plugins/helper/database/dbutil"
+)
+
+const defaultMysqlRevocationStmts = `
+	REVOKE ALL PRIVILEGES, GRANT OPTION FROM '{{name}}'@'%'; 
+	DROP USER '{{name}}'@'%'
+`
+const mySQLTypeName = "mysql"
+
+type MySQL struct {
+	connutil.ConnectionProducer
+	credsutil.CredentialsProducer
+}
+
+// New implements builtinplugins.BuiltinFactory
+func New() (interface{}, error) {
+	connProducer := &connutil.SQLConnectionProducer{}
+	connProducer.Type = mySQLTypeName
+
+	credsProducer := &credsutil.SQLCredentialsProducer{
+		DisplayNameLen: 4,
+		UsernameLen:    16,
```

Comment:
As of mysql 5.7.8 from 2015-08-03, the username length can be 32 chars. I think we should default to this because it lets us include more random UUID bits, more room for the display name, and more room for `vault_` in front. We could have a config option for the database be to override this value (um, right?). Or, even better would be if we can look up the version and do the right thing based on that.

This gets even messier, too:
MariaDB 5.5.31 supports up to 80 chars.
Aurora supports 16.
RDS supports 16.

We need a way to deal with this; being able to easily handle variations like this was one of the motivations for the combined database revamp in the first place :-)